code/modules/power/stirling.dm

```diff
@@ -0,0 +1,218 @@
+#define HEAT_TRANSFER 5000
+#define MAX_FREQUENCY 60
+#define MIN_DELTA_T 5
+
+/obj/machinery/atmospherics/binary/stirling
+	name = "stirling engine"
+	desc = "A mechanical stirling generator. It generates power dependent on the temperature differential between input gas and a radiative heatsink"
+	icon = 'icons/obj/power.dmi'
+	icon_state = "stirling"
+	density = 1
+	anchored = 1
+	layer = STRUCTURE_LAYER
+	use_power = POWER_USE_OFF
+	base_type = /obj/machinery/atmospherics/binary/stirling
+	construct_state = /decl/machine_construction/default/panel_closed
+	uncreated_component_parts = null
+	stat_immune = NOSCREEN | NOINPUT | NOPOWER // Mechanical machine and display.
+	connect_types = CONNECT_TYPE_REGULAR|CONNECT_TYPE_FUEL
+
+	var/obj/item/tank/stirling/inserted_cylinder
+
+	var/cycle_frequency = MAX_FREQUENCY/2
+
+	var/max_power = 100000
+	var/genlev = 0
+	var/lastgen = 0
+	var/skipped_cycle = FALSE
+
+	var/sound_id
+	var/datum/sound_token/sound_token
+
+/obj/machinery/atmospherics/binary/stirling/Destroy()
+	QDEL_NULL(inserted_cylinder)
+	QDEL_NULL(sound_token)
+	. = ..()
+
+/obj/machinery/atmospherics/binary/stirling/Process()
+	..()
+
+	// Some temperature equilibrium between the gas lines.
+	var/air1_eq = air1.get_thermal_energy_change(air2.temperature)
+	var/air2_eq = air2.get_thermal_energy_change(air1.temperature)
+
+	var/heat_transfer = clamp(HEAT_TRANSFER*(air1.temperature - air2.temperature), min(air1_eq, air2_eq), max(air1_eq, air2_eq))
+	if(heat_transfer)
+		air1.add_thermal_energy(-heat_transfer)
+		air2.add_thermal_energy(heat_transfer)
+
+	if(!istype(inserted_cylinder))
+		return
+
+	if(use_power != POWER_USE_ACTIVE)
+		update_icon()
+		update_sound()
```

Comment:
Do we really need to do this? Is it enough to do it once when power is toggled?

Reply:
Removed.